### DISCUSSION TITLE
watch seems to not work when i have two inputs and the help text is wrong it shows 0 characters

### DISCUSSION THREAD

**Discussion by genacatharine** [react-hook-form/react-hook-form#7668] [General] [answered]
hi i am trying to add an error if user goes above character limits in text field, it seems as thought the first text field works when i copy and paste too much text into it but the second field does not 
`  const titleValue = watch('title')
  const descriptionValue = watch('description')

 <TextField
                name='title'
                defaultValue=''
                label='Title*'
                control={control}
                css={{}}
                help={
                  errors.title?.type === 'required' || errors.title?.type === 'maxLength'
                    ? errors.title.message
                    : undefined
                }
                aria-label='title'
                rules={{
                  required: { value: true, message: 'Required' },
                  maxLength: {
                    value: 1024,
                    message: `${titleValue?.length}/1024 maximum character length exceeded`,
                  },
                }}
              />
            </ColumnLayout.Column>
            <ColumnLayout.Column span={3}>
              <ControlGroup label='Severity' labelPosition='top'>
                <Controller
                  control={control}
                  name='severity'
                  render={({ field }) => (
                    <Select
                      value={field.value}
                      onChange={(_, { value }) => field.onChange(value)}
                      aria-label='severity'
                    >
                      {severityOptions}
                    </Select>
                  )}
                />
              </ControlGroup>
            </ColumnLayout.Column>
          </ColumnLayout.Row>
          <ColumnLayout.Row>
            <ColumnLayout.Column span={12}>
              <TextField
                css={{}}
                defaultValue=''
                label='Description*'
                multiline
                aria-label='description'
                name='description'
                rules={{
                  required: { value: true, message: 'Required' },
                  maxLength: {
                    value: 4096,
                    message: `${descriptionValue?.length}/4096 maximum character length exceeded`,
                  },
                }}
                help={
                  errors.description?.type === 'required' || errors.description?.type === 'maxLength'
                    ? errors.description.message
                    : null
                }
                control={control}
              />

`

Textfield is `export const TextField: React.FC<TextFieldProps> = ({
  control,
  defaultValue,
  help,
  label,
  multiline,
  name,
  rowsMin,
  rules,
}) => {
  const { field, fieldState } = useController({ control, defaultValue, name, rules })

  return (
    <ControlGroup
      error={fieldState.invalid}
      label={label || capitalizeFirstLetter(name)}
      labelPosition='top'
      {...{ help }}
    >
      <Text error={fieldState.invalid} {...{ ...field, multiline, rowsMin }} />
    </ControlGroup>
  )
}
`

**Comment by bluebill1049** [comment]
can you plz reproduce that in a codesandbox?

**Reply by Moshyfawn** [reply]
Alright, I've reproduced it. It's not a MUI issue. It's watched values lagging behind one render. You can clearly see it if you paste "Title is very long" into the input and then paste it again the second time. The length of the subscribed input value is gonna be equal to the first time you've pasted the "Title is very long" string.

https://user-images.githubusercontent.com/16290753/151142340-96d2c749-1d27-4a09-be5c-748ccacf6fe9.mp4


[Here's a CSB](https://codesandbox.io/s/xenodochial-sid-p9fmq?file=/src/Demo.tsx). It's a demo with a controlled MUI `<TextField />`, but I've reproduced it with a native uncontrolled HTML input element.

**Reply by bluebill1049** [reply]
thanks @Moshyfawn this helps!

**Comment by bluebill1049** [comment, marked as answer]
it's not watching lag one render behind, validate function will also be updated in the render function. so if you want to have precise count down on the text, the error message needs to be put inside the render instead of the validate function.

**Reply by Moshyfawn** [reply]
So, you're saying, if you want to reference the current value of an input inside the error message, you can't use rules for that? Well, you can, but you gotta return some meaningful string to parse / dynamically select an error message string from some messages Map or object instead like
```js
// Rules
rules={{
  maxValue: {
    value: 10,
    message: 0
  }
}}
// Input error message
<p>{!!fieldState.error && errorMessages[fieldState.error.message] {field.value}}</p>
```

**Reply by bluebill1049** [reply]
```
-> render function
-> validate function (updated) // cached with old value from watch
-> render function 
```

<p>{!!fieldState.error && `custom error with length info ${title.length}`</p>

**Reply by Moshyfawn** [reply]
Yeah, but if I have length, required, regex etc. errors, it's gonna turn into switch statement party, right? Like
```
{fieldState.error?.type === "length" && `${field.value.length} is too long`}
{fieldState.error?.type === "regex" && `${field.value} is wrong`}
```

**Reply by bluebill1049** [reply]
no, not to do with validation rules. only the error message is affected. Validation is fixed at the render.

```
validate: (value) => {
  if (value < 3) // fixed
    return watchedValue // this will be one render behind because function closure
}
```

**Reply by Moshyfawn** [reply]
Ah, I think I understand where I went wrong with how to think about. Thank you @bluebill1049!
So, the solution is to do custom validate function like this
```js
rules={{
  validate: (value) => {
    if (value.length >= 10) {
      // this will give you the correct value for your error message
      return `Please, enter a title with less than 10 characters (${value.length}/10)`;
    }
  }
}}
```
[Example CSB](https://codesandbox.io/s/serene-bhabha-c3mou?file=/src/Demo.tsx)

**Reply by bluebill1049** [reply]
👍  or you can keep the message inside the render function as well, not as convenient.

**Reply by Moshyfawn** [reply]
> 👍 or you can keep the message inside the render function as well, not as convenient.

Yeah, I think, with smth like [i18n](https://react.i18next.com/) utils, it's easier to write your custom `validate` function due to things like [plurals](https://react.i18next.com/latest/trans-component#plural) factories to pass the value right into it.
If you just wanna have a reference in your error message, you can just do
```js
{!!fieldState.error && `custom error with length info ${watchedTitle.length}`
```
in your render function

**Reply by bluebill1049** [reply]
agrees 👍

**Reply by genacatharine** [reply]
thank you! I used your first solution 
`rules={{
  validate: (value) => {
    if (value.length >= 10) {
      // this will give you the correct value for your error message
      return `Please, enter a title with less than 10 characters (${value.length}/10)`;
    }
  }
}}`
